File: arrays/mergeAlternately.go
package main

import (
	"strings"
)

/*

link problem leetcode -> https://leetcode.com/problems/merge-strings-alternately/
time complexity  -> O(n)
space complexity -> O(n) 

*/

func mergeAlternately(word1 string, word2 string) string {
	var builder strings.Builder
	i := 0
	for i < len(word1) && i < len(word2) {
		builder.WriteString(string(word1[i]))
		builder.WriteString(string(word2[i]))
		i++
	}

	if i < len(word1) {
		builder.WriteString(string(word1[i:]))
	}

	if i < len(word2) {
		builder.WriteString(string(word2[i:]))
	}

	return builder.String()

}

File: arrays_and_strings/mergeAlternately.py
#Link problem: https://leetcode.com/problems/merge-strings-alternately/
#Time complexity: O(N), WHERE N = len(word1) + len(word2)
#Space complexity: O(N), WHERE N = len(word1) + len(word2)

class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        min_length = min(len(word1),len(word2))

        result = []

        for i in range(min_length):
            result.append(word1[i])
            result.append(word2[i])

        if len(word1) > min_length:
            result.append(word1[min_length:])
        elif len(word2) > min_length:
            result.append(word2[min_length:])

        return ''.join(result)


solution = Solution()
print(solution.mergeAlternately("abc","pqr"))
print(solution.mergeAlternately("ab","pqrs"))